cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogFolderCleaner.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.logging.clean;
+
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * LogFolderCleaner is responsible for cleaning up accumulated empty logs folders.
+ */
+public class LogFolderCleaner implements Runnable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogFolderCleaner.class);
+
+  private final Location logsDirectoryLocation;
+  private final int maxFolderCount;
+
+  private int count;
+
+  /**
+   * Parameterized constructor for LogFolderCleaner.
+   *
+   * @param logsDirectoryLocation Logs parent folder location.
+   * @param maxFolderCount Maximum number of folders to be cleaned up
+   */
+  public LogFolderCleaner(Location logsDirectoryLocation, int maxFolderCount) {
+    this.logsDirectoryLocation = logsDirectoryLocation;
+    this.maxFolderCount = maxFolderCount;
+  }
+
+  @Override
+  public void run() {
+    try {
+      LOG.debug("Starting log folder cleanup at timestamp {}", System.currentTimeMillis());
```

Comment:
I think this logic can be incorporated into the LogCleaner that runs periodically. The good thing about the LogCleaner is that it is aware of the timestamp as well, so that in every `run` cycle, it can also just delete all directories that are one interval older than the retention time.

Reply:
I believe the main challenge here is that the log storage does not know about log file name format, so it can't tell which ones are older than retention cycle. Having this understanding would be good as it would allow to remove also any files that are left behind for whatever reason. But it would also be somewhat more complicated

Reply:
Following changes are covered:
1. Merge LogCleaner and LogFolderCleaner.
2. Perform folder scans at every LogCleaner run to delete empty folders in batches of 100k.
3. Exclude folders for current date and previous date from traversal.
4. While there are more empty folders to be cleaned up, reschedule the next run at every hour. Otherwise schedule the next run at 24 hours.